### PR TITLE
fix main awaits

### PR DESCRIPTION
**Pull request by ksjogo**
Currently the main process didn't correctly call the await and was silently swallowing errors like the EACCESS ones from the bloodfever ci builds: https://github.com/KronbergerSpiele/bloodfever/runs/4633955575?check_suite_focus=true#step:6:292

Related to #51 (It isn't a fix for that, but now shows the correct error, which I currently think is unrelated)

I did add a lint rule to avoid these issues in the future.

Additionally use the promised version in the source compilation path to streamline error handling.

### CODE REVIEW

File: tests/project_tests.ts [hunk 1]
```diff
@@ -11,4 +11,7 @@ export const test = () => {
   })
 }
 
-test()
+test().catch((e) => {
```
Comment (johnfn) [inline tests/project_tests.ts:14]: dumb nit but I don't love .catch() mainly because people might thing it's the proper thing to do in other cases (when in fact you should use try/catch). We could extract this into a function... or just bump TS to ES2017 so we can use top level async/await :)

**Comment by johnfn** [comment]
Yea yea this looks good. Really just left some trivial nits but I'm willing to merge it whenever you want :)

File: tests/project_tests.ts (hunk 1 above)
Comment (ksjogo) [reply tests/project_tests.ts:14]: I think we already target ESNEXT, the issue is that we have a module:commonjs which doesn't have top level await.
Can switch them to immediately invoked ones, sure.

File: main.ts [hunk 2]
```diff
@@ -284,11 +284,19 @@ export const main = async (args: ParsedArgs) => {
...
   }
 
   showLoadingMessage("Compiling all source files", args)
-  project.compileAllSourceFiles()
+  try {
+    await project.compileAllSourceFiles()
+  } catch (e) {
+    // if in watch mode, try continuing
+    // in build mode, exit early with the error
+    if (args.buildOnly) {
```
[the remark this answers is not in the document]
Comment (ksjogo) [reply main.ts:296]: This is logged by the main(args) call in the same file.

File: main.ts [hunk 3]
```diff
@@ -317,6 +325,9 @@ if (!process.argv[1].includes("test")) {
   } else if (args.printVersion) {
     // Nothing to do; we already printed the version.
   } else {
-    main(args)
```
[the remark this answers is not in the document]
Comment (ksjogo) [reply main.ts:320]: Wrapped.

**Comment by ksjogo** [comment]
Updated.
I also converted some of the compilation to the async version to streamline the behaviour there.

File: main.ts [hunk 4]
```diff
@@ -297,7 +306,7 @@ export const main = async (args: ParsedArgs) => {
       true
     )
 
-    process.exit()
+    process.exit(hadErrors ? -1 : 0)
```
Comment (ksjogo) [inline main.ts:309]: The error behaviour here isn't great. We should probably print them all again.
But the global error file makes that a bit inconvenient. Should be another PR to rework that into project based error handling.

**Comment by johnfn** [comment]
Thanks!

**Comment by johnfn** [comment]
Thanks!
